### PR TITLE
Animated style issues

### PR DESCRIPTION
**Pull request by rbanerjee20**
### Clubhouse Ticket
[ch221](https://app.clubhouse.io/manim-dsl/story/221/null-pointer-exception-for-array-animation-styling)

### Description

Fixed several issues with animation styling - detailed in comments.

### Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue) :bug

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by mayanksharma3 — APPROVED** [review]

File: src/main/kotlin/com/manimdsl/linearrepresentation/LinearRepresentation.kt [hunk 1]
```diff
@@ -75,9 +76,10 @@ data class StackPopObject(
     }
 }
 
-data class ArrayElemAssignObject(val arrayIdent: String, val index: Int, val newElemValue: ExecValue) : ManimInstr {
+data class ArrayElemAssignObject(val arrayIdent: String, val index: Int, val newElemValue: ExecValue, val animatedStyle: AnimationProperties?) : ManimInstr {
```
Comment (rbanerjee20) [inline src/main/kotlin/com/manimdsl/linearrepresentation/LinearRepresentation.kt:79]: I realise this is quite tightly coupled and not using the `RestyleObject`, but that is owing to the current nature of our `replace_text` function (see line 82). Am welcome to any suggestions on how to get this working more cleanly

Comment (rbanerjee20) [reply src/main/kotlin/com/manimdsl/linearrepresentation/LinearRepresentation.kt:79]: Also see `rectangle.py`

File: src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt [hunk 2]
```diff
@@ -236,15 +249,32 @@ class VirtualMachine(
...
                     ArrayElemAssignObject(
                         (arrayValue.manimObject as ArrayStructure).ident,
                         index.value.toInt(),
-                        assignedValue
+                        assignedValue,
+                        arrayValue.animatedStyle
                     )
                 )
-                linearRepresentation.add(ArrayElemRestyle((arrayValue.manimObject as ArrayStructure).ident, listOf(index.value.toInt()), arrayValue.style))
+                if (arrayValue.animatedStyle != null) {
```
Comment (rbanerjee20) [inline src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt:269]: It should only be styling 'back to normal' if there was an animated style to start with, which is why I added this check

File: src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt [hunk 3]
```diff
@@ -459,10 +506,10 @@ class VirtualMachine(
             return when (node.type) {
                 is StackType -> {
                     val stackValue = StackValue(EmptyMObject, Stack())
-                    stackValue.style = stylesheet.getStyle(assignLHS.identifier, stackValue)
                     val initStructureIdent = variableNameGenerator.generateNameFromPrefix("stack")
                     dataStructureBoundaries[initStructureIdent] = TallBoundary()
                     stackValue.style = stylesheet.getStyle(assignLHS.identifier, stackValue)
+                    stackValue.animatedStyle = stylesheet.getAnimatedStyle(assignLHS.identifier, stackValue)
```
Comment (rbanerjee20) [inline src/main/kotlin/com/manimdsl/runtime/ASTExecutor.kt:512]: Added an `animatedStyle` property to both `StackValue` and `ArrayValue`, which uses the `getAnimatedStyle` function to merge the animated styles

File: src/main/kotlin/com/manimdsl/stylesheet/Stylesheet.kt [hunk 4]
```diff
@@ -77,14 +78,28 @@ class Stylesheet(private val stylesheetPath: String?, private val symbolTableVis
         val dataStructureStyle =
             stylesheet.dataStructures.getOrDefault(value.toString(), StyleProperties())
         val style = stylesheet.variables.getOrDefault(identifier, dataStructureStyle)
-        return style merge dataStructureStyle
+
+        val newStyle = style merge dataStructureStyle
+        val animatedStyle = getAnimatedStyle(identifier, value)
+        if (animatedStyle != null) {
+            if (animatedStyle.borderColor != null && newStyle.borderColor == null) {
```
Comment (rbanerjee20) [inline src/main/kotlin/com/manimdsl/stylesheet/Stylesheet.kt:85]: There was an issue with animations not switching back to the default WHITE when there was no textColor specified anywhere in the stylesheet. They would just stay the animation colour

**Review by iz0825 — COMMENTED** [review]
Looks good to me! Not really sure how we can reduce the coupling Rini mentioned, maybe @mayanksharma3 @claralebbos do you guys have any idea?

File: src/main/kotlin/com/manimdsl/linearrepresentation/LinearRepresentation.kt (hunk 1 above)
Comment (mayanksharma3) [reply src/main/kotlin/com/manimdsl/linearrepresentation/LinearRepresentation.kt:79]: Yeah I don't think there's a better way given we want to make the manim side as clean as possible (without unnecessary lines to fade to the color it already is)

**Comment by mayanksharma3** [comment]
> Looks good to me! Not really sure how we can reduce the coupling Rini mentioned, maybe @mayanksharma3 @claralebbos do you guys have any idea?

I think this looks good too, there's no real way round it I think, all I can think of is making a util function in manim to handle style changes, and so that is _always_ called but the color change only happens if it needs to. But that again isn't much better on the code gen side so I think this is good

**Review by mayanksharma3 — APPROVED** [review]

**Pull request merged**